% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/PrepareAnnotationRefseq.R
\name{PrepareAnnotationRefseq}
\alias{PrepareAnnotationRefseq}
\title{prepare annotation for Refseq}
\usage{
PrepareAnnotationRefseq(genome = "hg19", CDSfasta, pepfasta, annotation_path,
  dbsnp = NULL, transcript_ids = NULL, splice_matrix = FALSE,
  COSMIC = FALSE, local_cache_path = NULL, ...)
}
\arguments{
\item{genome}{specify the UCSC DB identifier (e.g. "hg19")}

\item{CDSfasta}{path to the fasta file of coding sequence.}

\item{pepfasta}{path to the fasta file of protein sequence, check 'introduction' for more detail.}

\item{annotation_path}{specify a folder to store all the annotations.}

\item{dbsnp}{specify a snp dataset to be used for the SNP annotation, default is NULL. (e.g. "snp135")}

\item{transcript_ids}{optionally, only retrieve transcript annotation data for the specified set of transcript ids. Default is NULL.}

\item{splice_matrix}{whether generate a known exon splice matrix from the annotation. this is not necessary if you don't want to analyse junction results, default is FALSE.}

\item{COSMIC}{whether to download COSMIC data, default is FALSE.}

\item{local_cache_path}{if non-NULL, refers to a directory where previously downloaded resources
(like protein coding sequences and COSMIC data) are cached so that the function can be re-run without
needing to download identical data again}

\item{...}{additional arguments}
}
\value{
several .RData file containing annotations needed for further analysis.
}
\description{
prepare the annotation for Refseq through UCSC table browser.
}
\examples{

transcript_ids <- c("NM_001126112", "NM_033360", "NR_073499", "NM_004448",
                    "NM_000179", "NR_029605", "NM_004333", "NM_001127511")
pepfasta <- system.file("extdata", "hg19/hg19_protein.fasta", package="customProDB")
CDSfasta <- system.file("extdata", "hg19/hg19_coding.fasta", package="customProDB")
cache_path <- system.file("extdata", "cache", package="customProDB")
annotation_path <- tempdir()
PrepareAnnotationRefseq(genome='hg19', CDSfasta, pepfasta, annotation_path,
            dbsnp=NULL, transcript_ids=transcript_ids,
            splice_matrix=FALSE, COSMIC=FALSE, local_cache_path=cache_path)

\dontrun{
dbkey = "hg38"
tempLocalCache = tempdir()
refseqTrack = ifelse(dbkey=="hg38", "refSeqComposite", "refGene")
codingFastaFilepath = paste0(tempLocalCache, "/", dbkey, ".cds.fa")
proteinFastaFilepath = paste0(tempLocalCache, "/", dbkey, ".protein.fa")

options(timeout=3600)
if (!file.exists(codingFastaFilepath)) {
  cat(paste("Downloading coding FASTA from:", ucscTableCodingFastaURL, "\\n"))
  download.file(getCodingFastaUrlFromUCSC(dbkey), codingFastaFilepath, quiet=T, mode='wb')
}

if (!file.exists(proteinFastaFilepath)) {
  cat(paste("Downloading protein FASTA from:", ucscTableProteinFastaURL, "\\n"))
  download.file(getProteinFastaUrlFromUCSC(dbkey), proteinFastaFilepath, quiet=T, mode='wb')
}

cat(paste("Preparing Refseq annotation files\\n"))
customProDB::PrepareAnnotationRefseq(dbkey, codingFastaFilepath, proteinFastaFilepath,
                                     annotation_path=".",
                                     dbsnp="snp146", COSMIC=FALSE,
                                     local_cache_path=tempLocalCache)
                                     
}
}
\author{
Xiaojing Wang
}
